### PR TITLE
[RFC] threads and channels

### PR DESCRIPTION
**Pull request by SamWilsn**
Here is another option for communication between the web server and the simulation that uses only threads and channels from the standard library.

### CODE REVIEW

File: notion-server/src/ethereum/mod.rs [hunk 1]
```diff
@@ -1,7 +1,140 @@
...
+        let (reply_send, reply_recv) = mpsc::sync_channel(1);
+
+        self.sender
+            .send(Operation::CreateExecutionEnvironment(args, reply_send))
+            .unwrap();
+
+        reply_recv.recv().unwrap()
+    }
+
+    pub fn join(self) -> Result<()> {
```
Comment (gjtrowbridge) [inline notion-server/src/ethereum/mod.rs:116]: What's the purpose of this method?  Just to stop the simulation?  If so, why not call it "stop"?

File: notion-server/src/ethereum/mod.rs [hunk 2]
```diff
@@ -1,7 +1,140 @@
...
+            }
+        };
+
+        drop(self);
+        handle.join().expect("simulation thread panicked")
+    }
+}
+
+#[derive(Debug)]
+struct EthereumSimulation {
```
Comment (gjtrowbridge) [inline notion-server/src/ethereum/mod.rs:137]: Were we to use two structs like this I'd push for something like `EthereumSimulation` and `EthereumSimulationInternal`.  Simulation and EthereumSimulation make it sound like they're different things, vs one just being a wrapper around the other.

**Review by gjtrowbridge — COMMENTED** [review]
Generally speaking I like the approach of using libraries that are extremely well-supported and well-documented (ie using standard library code as you've done here).  This seems like a great approach, and based on a quick glance, should allow us to craft the same external interface and same modularization that we'd get with the other approaches.  I think this is where I'd lean, though definitely open to discussing in purpose.

Thanks again for all the time and work in putting these together, will take a look at #3 now...

File: notion-server/src/ethereum/mod.rs (hunk 1 above)
Comment (SamWilsn) [reply notion-server/src/ethereum/mod.rs:116]: It's a combination of "stop the simulation" and "wait for it to finish".

`stop` is fine too.

File: notion-server/src/ethereum/mod.rs (hunk 2 above)
Comment (SamWilsn) [reply notion-server/src/ethereum/mod.rs:137]: Since it isn't a public item, I don't care too much about the naming. I'm okay with whatever :)

**Pull request closed without merging**